ldms/src/core/ldms.c

```diff
@@ -309,9 +309,11 @@ struct get_set_names_arg {
  * }
  */
 
-static const char *perm_string(uint32_t perm)
+#define PERM_BUF_SZ 16
+typedef char(*perm_buf)[PERM_BUF_SZ];
+static const char *perm_string(uint32_t perm, perm_buf pstr)
```

Comment:
Pass in a char *. This function is called in one place and always consumes exactly 10B.

Reply:
@tom95858 You've hit on the core of the problem which is that some simple formatting was extracted into single-use functions; this introduced the races and the opportunity to mismatch buffer sizes.
How about I just move the formatting back where it belongs in __ldms_format_set_meta_as_json?

Reply:
That is fine with me.